src/index.js

```diff
@@ -1,198 +1,48 @@
 'use strict'
 
-const debug = require('debug')
-const log = debug('libp2p:tcp')
-const tcp = require('net')
-const multiaddr = require('multiaddr')
-const Address6 = require('ip-address').Address6
+const net = require('net')
+const toPull = require('stream-to-pull-stream')
 const mafmt = require('mafmt')
-// const parallel = require('run-parallel')
 const contains = require('lodash.contains')
-const os = require('os')
+const isFunction = require('lodash.isfunction')
 const Connection = require('interface-connection').Connection
+const debug = require('debug')
+const log = debug('libp2p:tcp')
 
-exports = module.exports = TCP
-
-const IPFS_CODE = 421
-const CLOSE_TIMEOUT = 2000
-
-function TCP () {
-  if (!(this instanceof TCP)) {
-    return new TCP()
-  }
+const createListener = require('./listener')
 
-  this.dial = function (ma, options, callback) {
-    if (typeof options === 'function') {
-      callback = options
+module.exports = class TCP {
+  dial (ma, options, cb) {
+    if (isFunction(options)) {
+      cb = options
       options = {}
     }
 
-    if (!callback) {
-      callback = function noop () {}
+    if (!cb) {
+      cb = () => {}
     }
 
-    const socket = tcp.connect(ma.toOptions())
-    const conn = new Connection(socket)
-
-    socket.on('timeout', () => {
-      conn.emit('timeout')
-    })
-
-    socket.once('error', (err) => {
-      callback(err)
-    })
-
-    socket.on('connect', () => {
-      callback(null, conn)
-      conn.emit('connect')
-    })
+    const cOpts = ma.toOptions()
+    log('Connecting to %s %s', cOpts.port, cOpts.host)
+    const socket = toPull.duplex(net.connect(cOpts, cb))
```

Comment:
update: 
- We don't need the `connect` event, as we can use the callback for that
- [ ] We do need to listen for the `timeout` event, so that a cb(new Error('timeout') is emitted. Let's add a test for this.